Add raw mmap

diwic opened the pull request:
Hi!

In order to use the memmap2 crate for my [IPC experiments](https://github.com/diwic/shmem-ipc), I had to create a raw mmap which just exposes a pointer and a length, no slices.

Due to Rust's aliasing rules, creating slices to memory can be unsound if more than one process can write to the same memory area. (The IPC protocol can dictate which parts of the memory that can be accessed in what way.)

RazrFalcon commented on src/lib.rs:
```diff
@@ -468,6 +479,39 @@ impl fmt::Debug for Mmap {
+impl MmapRaw {
+    /// A raw pointer to the interior.
+    ///
+    /// ## Safety
```
One `#` is enough.

RazrFalcon commented on src/lib.rs:
```diff
@@ -468,6 +479,39 @@ impl fmt::Debug for Mmap {
+    /// truncated since the memory map was created.
+    #[inline]
+    pub fn as_ptr(&self) -> *mut u8 { self.inner.ptr() as _ }
+    /// The length in bytes of the memory map.
```
New line.

RazrFalcon commented on src/lib.rs:
```diff
@@ -468,6 +479,39 @@ impl fmt::Debug for Mmap {
+    /// To safely dereference this pointer, you need to make sure that the file has not been
+    /// truncated since the memory map was created.
+    #[inline]
+    pub fn as_ptr(&self) -> *mut u8 { self.inner.ptr() as _ }
```
This should return `*const u8`. And `*mut u8` should be under `as_mut_ptr`.

diwic commented:
Thanks for the quick review! I've pushed another commit that addresses your feedback. Let me know if you have any further thoughts.

After merging, I would appreciate a crates.io release (so I can publish my crate on crates.io). Thanks in advance!

RazrFalcon commented on src/lib.rs:
```diff
@@ -468,6 +479,62 @@ impl fmt::Debug for Mmap {
+        MmapOptions::new().map_raw(file)
+    }
+
+    /// A raw mutable pointer to the interior.
```
Rename to _Returns a raw pointer to the memory mapped file._ To match [slice::as_ptr](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.as_ptr)

RazrFalcon commented on src/lib.rs:
```diff
@@ -468,6 +479,62 @@ impl fmt::Debug for Mmap {
+    #[inline]
+    pub fn as_ptr(&self) -> *const u8 { self.inner.ptr() }
+
+    /// A raw mutable pointer to the interior.
```
Rename to _Returns an unsafe mutable pointer to the memory mapped file._

RazrFalcon commented on src/lib.rs:
```diff
@@ -468,6 +479,62 @@ impl fmt::Debug for Mmap {
+impl MmapRaw {
+    /// Creates a writeable memory map backed by a file.
+    ///
+    /// This is equivalent to calling `MmapOptions::new().map_mut(file)`.
```
Change to `MmapOptions::new().map_raw(file)`

RazrFalcon commented on src/lib.rs:
```diff
@@ -468,6 +479,62 @@ impl fmt::Debug for Mmap {
+    #[inline]
+    pub fn as_mut_ptr(&self) -> *mut u8 { self.inner.ptr() as _ }
+
+    /// The length in bytes of the memory map.
```
Rename to _Returns the length in bytes of the memory map._

RazrFalcon commented on src/lib.rs:
```diff
@@ -468,6 +479,62 @@ impl fmt::Debug for Mmap {
+
+impl fmt::Debug for MmapRaw {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("Mmap")
```
`Mmap` -> `MmapRaw`

diwic commented:
Alright, second round fixed now. Thanks for the quick review!

RazrFalcon commented on src/lib.rs:
````diff
@@ -468,6 +479,62 @@ impl fmt::Debug for Mmap {
+    ///
+    /// This method returns an error when the underlying system call fails, which can happen for a
+    /// variety of reasons, such as when the file is not open with read and write permissions.
+    /// ```
````
Remove this line.

RazrFalcon commented:
Almost there.

diwic commented:
Fixed and squashed into one commit. Sorry about the sloppiness - comments are the hardest to get right, because the compiler does not check them for you ;-)

RazrFalcon commented:
No problem.

RazrFalcon commented:
Published.

diwic commented:
Thanks!
